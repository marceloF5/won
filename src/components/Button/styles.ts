import styled, { css, DefaultTheme } from 'styled-components'
import { darken } from 'polished'
import { ButtonProps } from '.'

export type WrapperProps = { hasIcon: boolean } & Pick<
    ButtonProps,
    'fullWidth' | 'minimal' | 'size'
>

const wrapperModifiers = {
    fullWidth: () => css`
        width: 100%;
    `,

    small: (theme: DefaultTheme) => css`
        height: 3rem;
        font-size: ${theme.font.sizes.xsmall};
    `,

    medium: (theme: DefaultTheme) => css`
        height: 4rem;
        font-size: ${theme.font.sizes.small};
        padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
    `,

    large: (theme: DefaultTheme) => css`
        height: 5rem;
        font-size: ${theme.font.sizes.medium};
        padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
    `,

    minimal: (theme: DefaultTheme) => css`
        background: none;
        color: ${theme.colors.primary};

        &:hover {
            color: ${darken(0.1, theme.colors.primary)};
        }
    `,

    withIcon: (theme: DefaultTheme) => css`
        svg {
            width: 1.5rem;

            & + span {
                margin-left: ${theme.spacings.xxsmall};
            }
        }
    `
}

export const Wrapper = styled.button<WrapperProps>`
    ${({ theme, fullWidth, hasIcon, minimal, size }) => css`
        display: inline-flex;
        align-items: center;
        justify-content: center;
        color: ${theme.colors.white};
        cursor: pointer;
        background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);
        border: 0;
        border-radius: ${theme.border.radius};
        padding: ${theme.spacings.xxsmall};
        text-decoration: none;

        &:hover {
            background: ${
                minimal
                    ? 'none'
                    : `linear-gradient(180deg, #e35565 0%, #d958a6 50%)`
            };
        }

        ${!!fullWidth && wrapperModifiers.fullWidth()}
        ${!!hasIcon && wrapperModifiers.withIcon(theme)}
        ${!!minimal && wrapperModifiers.minimal(theme)}
        ${!!size && wrapperModifiers[size](theme)};
    `}
`
