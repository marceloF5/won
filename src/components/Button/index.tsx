import * as S from './styles'

export type ButtonProps = {
    children?: React.ReactNode
    fullWidth?: boolean
    icon?: JSX.Element
    onClick?: () => (event: React.MouseEvent<HTMLButtonElement>) => void
    size?: 'small' | 'medium' | 'large'
}

const Button = ({
    children,
    fullWidth = false,
    icon,
    size = 'medium',
    ...otherProps
}: ButtonProps) => (
    <S.Wrapper
        fullWidth={fullWidth}
        hasIcon={!!icon}
        size={size}
        {...otherProps}
    >
        {!!icon && icon}
        {!!children && <span>{children}</span>}
    </S.Wrapper>
)

export default Button
