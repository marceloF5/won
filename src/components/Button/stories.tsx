import { Story, Meta } from '@storybook/react/types-6-0'
import { AddShoppingCart } from '@styled-icons/material-outlined'
import React from 'react'

import Button from '.'

export default {
    title: 'Button',
    component: Button,
    argTypes: {
        children: {
            type: 'string'
        },
        icon: {
            type: ''
        }
    }
} as Meta

export const Default: Story = (args) => <Button {...args} />
export const withIcon: Story = (args) => <Button {...args} />

Default.args = {
    children: 'Buy now'
}

withIcon.args = {
    size: 'small',
    children: 'Buy now',
    icon: <AddShoppingCart />
}
