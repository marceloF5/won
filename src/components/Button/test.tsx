import { screen } from '@testing-library/react'
import { AddShoppingCart } from '@styled-icons/material-outlined'
import React from 'react'
import { renderWithTheme } from 'utils/tests/helpers'

import Button from '.'

describe('<Button />', () => {
    it('should render default component corretly', () => {
        const { container } = renderWithTheme(<Button>Buy now</Button>)

        expect(container).toMatchSnapshot()
    })

    it('should render button small size', () => {
        renderWithTheme(<Button size={'small'}>Buy now</Button>)

        expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
            height: '3rem',
            'font-size': '1.2rem'
        })
    })

    it('should render button medium by default', () => {
        renderWithTheme(<Button size="medium">Buy now</Button>)

        expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
            height: '4rem',
            'font-size': '1.4rem',
            padding: '0.8rem 3.2rem'
        })
    })

    it('should render button large size', () => {
        renderWithTheme(<Button size={'large'}>Buy now</Button>)

        expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
            height: '5rem',
            'font-size': '1.6rem',
            padding: '0.8rem 4.8rem'
        })
    })

    it('should render a fullWidth button version', () => {
        renderWithTheme(<Button fullWidth>Buy now</Button>)

        expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
            width: '100%'
        })
    })

    it('should render a icon button version', () => {
        renderWithTheme(
            <Button icon={<AddShoppingCart data-testid="icon" />}>
                Buy now
            </Button>
        )

        expect(screen.getByText(/buy now/i)).toBeInTheDocument()
        expect(screen.getByTestId(/icon/i)).toBeInTheDocument()
    })

    it('should render a minimal button version', () => {
        renderWithTheme(
            <Button icon={<AddShoppingCart data-testid="icon" />} minimal>
                Buy now
            </Button>
        )

        expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
            background: 'none',
            color: '#F231A5'
        })

        expect(
            screen.getByRole('button', { name: /buy now/i })
        ).toHaveStyleRule('background', 'none', { modifier: ':hover' })
    })

    it('should render button as link', () => {
        renderWithTheme(
            <Button as="a" href="/link">
                Buy now
            </Button>
        )

        expect(screen.getByRole('link', { name: /buy now/i })).toHaveAttribute(
            'href',
            '/link'
        )
    })

    it('should render a disabled button', () => {
        renderWithTheme(<Button disabled>Buy now</Button>)

        expect(
            screen.getByRole('button', { name: /buy now/i })
        ).toHaveStyleRule('cursor', 'not-allowed', { modifier: ':disabled' })
    })
})
