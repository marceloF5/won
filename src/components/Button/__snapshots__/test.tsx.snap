// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render default component corretly 1`] = `
.c0 {
  color: #fafafa;
  cursor: pointer;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  border: 0;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

<div>
  <button
    class="c0"
  >
    <span>
      Buy now
    </span>
  </button>
</div>
`
